python/fastfec.py

```diff
@@ -63,79 +262,87 @@ def init_lib(self):
-         make_dirs = arg.lower() != 'false'
-   input_file = f'{os.path.join(input_directory, filing_id)}.fec'
-   output_directory = os.path.join(output_directory, filing_id)
-   print(f'Filing ID is {filing_id}')
-   print(f'Input file is {input_file}')
-   print(f'Output directory is {output_directory}')
-   print(f'Make parent directories: {make_dirs}')
-
-   f = open(
-     input_file, mode='rb', transport_params=dict(buffer_size=TRANSFER_BUFFER_SIZE))
-   fast_fec = FastFEC(f, output_directory, use_compression=use_compression, make_dirs=make_dirs)
-   fast_fec.parse()
-
-   # Free memory
-   fast_fec.free()
-   f.close()
-
-if __name__ == "__main__":
-   main(sys.argv[1:])
+        return [CUSTOM_WRITE(write_callback), free_file_descriptors]
+
+    def __set_read_callback(self, file_handle):
+        self.buffer_read_fn = BUFFER_READ(
+            make_read_buffer(file_handle))
+
+    def __provide_line_callback(self, queue):
+        # Initialize parsing cache
+        line_cache = LineCache()
+
+        def line_callback(filename, line, types):
+            # Somehow yield in parent function
+            if filename == line_cache.last_filename:
+                queue.put((filename.decode('utf8'), line_result(line_cache.last_headers,
+                                                                csv_parse(line.decode('utf8')), types)))
+                queue.join()
+            else:
+                # Grab the headers from the cache if possible
+                headers = line_cache.headers.get(filename)
+                first_line = False
+                if not headers:
+                    # The headers have not yet encountered. They
+                    # are always in the first line, so this line
+                    # will contain them.
+                    line_cache.headers[filename] = csv_parse(
+                        line.decode('utf8'))
+                    headers = line_cache.headers[filename]
+                    first_line = True
+                line_cache.last_filename = filename
+                line_cache.last_headers = headers
+                if not first_line:
+                    # Format the result and return it (if not a header)
+                    queue.put((filename.decode('utf8'), line_result(headers, csv_parse(
+                        line.decode('utf8')), types)))
+                    queue.join()
+        return line_callback
+
+    def __free_fec_context(self):
+        if self.fec_context:
+            self.libfastfec.freeFecContext(self.fec_context)
+            self.fec_context = None
+
+
+@ contextlib.contextmanager
+def FastFEC():
+    """A convenience method to run fastfec and free memory afterwards
+
+    Usage:
+
+        with FastFEC() as fastfec:
+            # Running fastfec parse methods
+            # The memory is freed afterwards, so no need to ever call
+            # fastfec.free()
+            ...
+    """
+    instance = LibFastFEC()
+    yield instance
+    instance.free()
```

Comment:
Not sure what this is doing bc I haven't seen `yield` before

Reply:
Yield just returns a generator